### ISSUE TITLE
[Test Failure] NtpOSGiTest fails randomly on my local machine during mvn clean install

### ISSUE THREAD

**Issue by triller-telekom** [eclipse/smarthome#3231]
Sometimes the test runs fine and sometimes it fails, I cannot really specify when. But it just happend twice after running "mvn clean install" -> reboot laptop -> "mvn clean install".

I am running Arch Linux on my machine.

testDateTimeChannelWithUnknownHost(org.eclipse.smarthome.binding.ntp.test.NtpOSGiTest)  Time elapsed: 10.182 sec  <<< FAILURE!
java.lang.AssertionError: The thing status was not communication error
Expected: is <COMMUNICATION_ERROR>
     but: was <NONE>
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
        at org.junit.Assert.assertThat(Assert.java:865)
        at org.eclipse.smarthome.binding.ntp.test.NtpOSGiTest.lambda$4(NtpOSGiTest.java:489)

**Comment by sjka** [comment]
I just had a closer look at the debug output of the test execution on your machine. This is how it looks like on _my_ machine:

```
14:19:47.192 [safeCall-1] DEBUG o.e.s.binding.ntp.handler.NtpHandler - Initialized NTP handler 'ntp:ntp:testThingId' with configuration: host 'wrong.hostname', refresh interval 60, timezone sun.util.calendar.ZoneInfo[id="Europe/Helsinki",offset=7200000,dstSavings=3600000,useDaylight=true,transitions=118,lastRule=java.util.SimpleTimeZone[id=Europe/Helsinki,offset=7200000,dstSavings=3600000,useDaylight=true,startYear=0,startMode=2,startMonth=2,startDay=-1,startDayOfWeek=1,startTime=3600000,startTimeMode=2,endMode=2,endMonth=9,endDay=-1,endDayOfWeek=1,endTime=3600000,endTimeMode=2]], locale en_US.
14:19:47.193 [ESH-thingHandler-3] WARN  o.e.s.binding.ntp.handler.NtpHandler - ntp:ntp:testThingId the given hostname 'wrong.hostname' of the timeserver is unknown -> returning current sytem time instead.
```

So it takes just a really short moment to realize that the host cannot be resolved. However, on your machine it behaves a little bit different:

```
14:31:30.124 [safeCall-1] DEBUG o.e.s.binding.ntp.handler.NtpHandler - Initialized NTP handler 'ntp:ntp:testThingId' with configuration: host 'wrong.hostname', refresh interval 60, timezone 
14:31:45.235 [ESH-thingHandler-3] WARN  o.e.s.binding.ntp.handler.NtpHandler - ntp:ntp:testThingId couldn't establish network connection [host 'wrong.hostname'] -> returning current sytem time instead.
```

Two things to notice:

1. It takes about 15 seconds to fail. Which is longer than the test timeout.
2. The log message is a bit different. 

When looking at the [NtpHandler](https://github.com/eclipse/smarthome/blob/master/extensions/binding/org.eclipse.smarthome.binding.ntp/src/main/java/org/eclipse/smarthome/binding/ntp/handler/NtpHandler.java#L232) and its log messages it means it did not fail with an `UnknownHostException` but with an `IOException`  instead. For whatever reasons it must have been able to resolve "wrong.hostname" to something sensible, tried to connect there but obviously failed.

I'm pretty sure this was not the intention of the test, although the resulting status would have been the same. So increasing the timeout in the test definitely _would_ help, but I rather suggest to keep it as it is since I would expect a DNS resolution error on any other machine.

Therefore I suggest to close this issue, being a problem which appears only on your specific linux configuration.

**Comment by triller-telekom** [comment]
Some more observations show that "random" is not really random here. It turns out that this test actually connects to REAL ntp servers, i.e. 0.pool.ntp.org which is bad practice for an integration test that runs ob every build on our build server.

The specific test that failed on my machine was because my machine could resolve "wrong.hostname". Thanks to the provider T-Online which has configured their DNS in such a way that it returns a valid IP for EVERY hostname. This is a comfort feature to redirect customers in the browser to a portal website saying "you might have a typo in your address".

Long story short: The dependency to any external ntp server should be removed from this test and we have to use a different DNS server :)

**Comment by maggu2810** [comment]
I agree that a test itself shouldn't depend on an internet connection.
We should consider how a test of a NTP client should be written do work without an internet connection.
Should we "imitate" a server and just offer the correct bytes that would be received by a real server?

The "feature" to return an IP address for a non-resolvable DNS entry is IMHO not a feature. If there is no DNS record, then I would expect a technical correct reply that it cannot be resolved.
But I never analyzed the DNS response myself, just realized the wired behavior.
Thanks to Google there are other well known nameservers that can I use at home.

**Comment by sjka** [comment]
Right, I think we all agree that the weird behavior of the DNS servers offered by a popular telco indeed should not bother us here. So let's ignore this point here.

Nevertheless, the dependency to the external system remains more than valid! (I just ran into another remote-communication related issue which randomly fails these test...). So as much as I love these tests for the fact that they are written in plain Java 8, I would like to propose disabling them as a first step until we got them fixed by a local DNS fake server. Wdyt?

**Comment by maggu2810** [comment]
> until we got them fixed by a local DNS fake server

DNS? I assume you mean NTP. Correct?

**Comment by sjka** [comment]
yes, NTP of course 😆

**Comment by svilenvul** [comment]
As you have discussed above, I have opened PR that will ignore the tests until we add a local fake NTP server.

**Comment by sjka** [comment]
just FTR:
The idea mentioned in #3280 is to...
> introduce a simple NTP server e.g. this one is available using the Apache License, Version 2.0: https://commons.apache.org/proper/commons-net//examples/ntp/SimpleNTPServer.java

**Comment by svilenvul** [comment]
Could someone open a CQ for the SimpleNTPServer ?

**Comment by sjka** [comment]
I just had a close look...

We already have a CQ for apache commons net. It includes the `NtpV3Packet` which is used in the `handlePacket()` method and does the real magic. For the matter of the test cases we would probably adapt this method to our needs anyway (especially as it has build-errors!). The rest of the example class is about opening and shutting down a socket.

Are you sure you want to include it as it is as third-party code? I would actually think that with the help of `NtpV3Packet`/`NtpV3Impl` it would be easiest to write our own simple server implementation with just a few lines of code... What do you think?

**Comment by svilenvul** [comment]
Great, thanks for looking at this.

Yes, you are absolutely right, in this case writing a simple server implementation will be the easiest option.
